feat: add a fuse to speed up the case where no resource is present

RaisinTen opened the pull request:
Fixes: https://github.com/postmanlabs/postject/issues/58
Signed-off-by: Darshan Sen <raisinten@gmail.com>

RaisinTen commented on src/api.js:
```diff
@@ -112,8 +112,34 @@ async function inject(filename, resourceName, resourceData, options) {
+  const firstSentinel = buffer.indexOf(SENTINEL);
+
+  if (firstSentinel === -1) {
+    throw new Error(`Could not find the sentinel ${SENTINEL} in the binary`);
```
CI is failing because this error is getting thrown. It happens because the Node.js binary we are testing against doesn't do `#include "postject-api.h"` which is supposed to embed the sentinel string.

Suggestions on what we should do here?

jviotti replied:
Postject is supposed to be completely generic and independent from i.e. Node.js or a specific application. Can we run the tests not against Node.js specifically, but mainly against a dummy program we compile from source during the testing phase?

RaisinTen replied:
@jviotti Are you suggesting that we should remove this whole thing - https://github.com/postmanlabs/postject/blob/30e9a23044f71ca16f9345e94266906acb931d37/test/cli.mjs#L150-L233?

jviotti replied:
Yeah, I think it makes sense. It feels that such test should be part of Node.js or our SEA CLI instead?

RaisinTen replied:
Cool, rebased. Tests pass locally now but we would need to reconfigure CircleCI to get a green CI. PTAL.

jviotti commented on postject-api.h:
```diff
@@ -1,6 +1,7 @@
 #ifndef POSTJECT_API_H_
 #define POSTJECT_API_H_
 
+#include <stdbool.h>
```
Ah, good old `stdbool` :P I had completely forgotten about this

robertgzr reviewed (commented):
since we're trying to enable others to implement either side of the process usign tools that are not postject, we should document this.

using [rfc lanuage](https://www.rfc-editor.org/rfc/rfc2119), is this going to be a `MAY`? so extract implementation have to work without it, but have to consider it when it is present?

tony-go commented on postject-api.h:
```diff
@@ -33,6 +34,11 @@ inline void postject_options_init(struct postject_options* options) {
 
+static inline bool postject_has_resource() {
+  const char* sentinel = "POSTJECT_SENTINEL_fce680ab2cc467b6e072b8b5df1996b2:0";
+  return sentinel[strlen(sentinel) - 1] == '1';
```
Is that won't always return `false`?

RaisinTen replied:
nope because if you look at the change in `src/api.js`, we are changing the last character in the sentinel string from `0` to `1`

dsanders11 commented on postject-api.h:
```diff
@@ -33,6 +34,11 @@ inline void postject_options_init(struct postject_options* options) {
   options->pe_resource_name = NULL;
 }
 
+static inline bool postject_has_resource() {
+  const char* sentinel = "POSTJECT_SENTINEL_fce680ab2cc467b6e072b8b5df1996b2:0";
+  return sentinel[strlen(sentinel) - 1] == '1';
+}
+
```
This implementation will be optimized out by the compiler - `sentinel` is a local value to `project_has_resource` so the compiler could decide it's a constant false condition and not actually evaluate it at run-time.

Check it in godbolt.org: https://godbolt.org/z/rPKd1sEWe. If you you add `-O3` as a compiler option (top of the right panel), then you can see the generated code collapses to be nothing, since it is a constant false condition.

I think something like this could work (unfortunately can't use `strlen` on a volatile):

```C
static inline bool postject_has_resource() {
  static const volatile char* sentinel = "POSTJECT_SENTINEL_fce680ab2cc467b6e072b8b5df1996b2:0";
  return sentinel[51] == '1';
}
```

dsanders11 commented on test/test.cpp:
```diff
@@ -1,15 +1,19 @@
+  size_t size = 0;
 
-  if (ptr && size > 0) {
+  if (postject_has_resource() == true) {
```
Nit: don't need the `== true`, same with above.

dsanders11 commented on test/test.cpp:
```diff
@@ -1,15 +1,19 @@
-  if (ptr && size > 0) {
+  if (postject_has_resource() == true) {
+    const void* ptr = postject_find_resource("foobar", &size, nullptr);
+    assert(ptr && size > 0);
```
I think this should be done with an if statement rather than an `assert`. We probably should be testing the final binary in both debug builds and release builds, and the `assert` will be removed when `-DNDEBUG` is set in a release build.

dsanders11 commented:
> using [rfc lanuage](https://www.rfc-editor.org/rfc/rfc2119), is this going to be a MAY? so extract implementation have to work without it, but have to consider it when it is present?

I think the other way around - if the run-time implementation (`postject-api.h` for the reference one) chooses to include it, the injection side should flip the value to indicate a resource is present. If the run-time implementation does not include it, the injector should be OK with it not being there and do nothing. I don't necessarily feel strongly one way or the other on it being optional, maybe leaning toward it would be nice if it was optional.

dsanders11 reviewed (changes requested):
Could the commit message be tweaked? This doesn't speed up resource access, it speeds up the case where there is no resource present. If anything it slows down resource access with an extra check. 😄

Also, I think it would be nice if the name of the sentinel was an option (with default value) for the injection side. I tried to make all references to the name Postject configurable so any user could customize it for various reasons.

ovflowd commented:
@RaisinTen any update here? 👀

RaisinTen commented:
@ovflowd I was thinking about addressing the remaining comments and landing once CI starts running on this repo again, see https://github.com/nodejs/postject/issues/62.

RaisinTen commented:
I've imported this patch in https://github.com/nodejs/node/pull/45038 and there are no related errors. PTAL everyone!

dsanders11 reviewed (approved):
Overall LGTM, although I think it would be more ergonomic if the `:0` suffix was internalized as an implementation detail and only `POSTJECT_SENTINEL_fce680ab2cc467b6e072b8b5df1996b2` was exposed to users. Then there wouldn't need to be a validation check for the CLI option - any string value would be valid. Would also help prevent issues since there is no validation on the macro side.

I believe this could be accomplished with in `postject-api.h` with something like:

```c
static inline bool postject_has_resource() {
  static const volatile char* sentinel = POSTJECT_SENTINEL_FUSE ":0";
  return sentinel[sizeof(POSTJECT_SENTINEL_FUSE)] == '1';
}
```

RaisinTen commented on src/cli.js:
```diff
@@ -52,6 +53,11 @@ if (require.main === module) {
     )
+    .option(
+      "--sentinel-fuse <sentinel_fuse>",
+      "Sentinel fuse",
```
Done! I'm open to better names if you have a suggestion

dsanders11 reviewed (commented):
@RaisinTen, changes look good (thanks!), but the tests still have `:0` in their usages. Once that's cleaned up I'm good to merge.

dsanders11 commented:
@RaisinTen, sounds good to me. Merging this PR is blocked by the old CircleCI status check being required, and I don't have perms to remove (or override) that check so I'll defer to you on merging, but LGTM. 👍
